Add Discord page

s-ben opened the pull request:
This PR fleshes out a first draft of the Discord plugin. 

I used the Discourse plugin page as a template. Instructions on configuring the bot and emoji weights are basically a rewrite of the Discord section in the [template-instance README](https://github.com/sourcecred/template-instance).

I did my best to fill out all I could, but there are definitely some holes due to not knowing some of the technicals. Have left comments pointing out holes or asking for review where guessing.

hammadj commented on docs/setup/installation.md:
```diff
@@ -18,8 +20,127 @@ If you want to build SourceCred from source (for instance to get bleeding edge f
 If you don't have it, install [yarn](https://classic.yarnpkg.com/lang/en/). Then run yarn to install SourceCred and its dependencies.
 
 `yarn` 
+<<<<<<< HEAD
```
Seems like there's an unresolved merge conflict here. Did you mean to edit this page?

hammadj commented on docs/setup/installation.md:
```diff
@@ -4,7 +4,9 @@ title: Installation
-If you want to build SourceCred from source (for instance to get bleeding edge features) or to develop on it, see the [Development Setup instructions](https://github.com/sourcecred/sourcecred#development-setup) on GitHub.
+If you want to build SourceCred from source (for instance to get bleeding edge features) or to develop on it, see the [Development Setup instructions](#development-setup).
+
+You can also run SourceCred from a [Docker image](#using-docker)  (useful for automated setups and canonical instances). 
```
We wont be supporting the docker image anymore since we have the NPM package now

s-ben replied:
Think this was an artifact of my git sloppiness, but removed here anyway. Don't think it warrants a separate PR worth the same amount of Cred as this whole page :)

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+- **User**:
+
+A Discord user account, e.g. [@decentralion {SB: correct link?}](). User accounts do not mint Cred, so setting a node weight would have no effect. Using the identity plugin, it's possible to "collapse" user nodes with other identity nodes into a single, canonical identity. For example, if a contributor had a Discord user account and a GitHub account, then the identity plugin can collapse those id ...
```
Maybe "merge" instead of "collapse" since thats the wording we use in the code / UI?

s-ben replied:
Merge is better. Going with that.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+A Discord user account, e.g. [@decentralion {SB: correct link?}](). User accounts do not mint Cred, so setting a node weight would have no effect. Using the identity plugin, it's possible to "collapse" user nodes with other identity nodes into a single, canonical identity. For example, if a contributor had a Discord user account and a GitHub account, then the identity plugin can collapse those id ...
+
+Users are connected to messages they author, to messages they react to, and to messages that mention them {SB: is this true?}.
```
Yes its true

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+- **Guild Member**:
+
+A Discord user account, e.g. [@decentralion {SB: correct?}](), that has special powers. E.g. their reactions mint more Cred {SB: totally making this up here. Is this part of beta launch, or just a hidden alpha feature for MetaGame?}
```
I would just take this out since it's not implemented right now and is not clear how exactly we will make it work. Even then, it wont be an actual node type.

s-ben replied:
Removed

s-ben replied:
Removed.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+A Discord user account, e.g. [@decentralion {SB: correct?}](), that has special powers. E.g. their reactions mint more Cred {SB: totally making this up here. Is this part of beta launch, or just a hidden alpha feature for MetaGame?}
+
+- **Bot**:
+
+{SB: below is copied from GitHub page. Still true for Discord? Do we support Discord bots?}
+A Discord user account that has been explicitly marked as a bot, via inclusion in bots.js. This is useful so we can filter out bot accounts from receiving grain or showing up in the Cred rankings.
+
+Bots have the same connections as users.
```
This isnt an actual node type, and AFAIK we dont support ignoring bots like that yet in discord. Can probably take this section out for now since we will likely have a specific way to ignore certain nodes in the future.

s-ben replied:
Removed.

s-ben replied:
Removed.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+- **Channel**:
+
+A Discord channel, e.g. [#start-here](https://discord.gg/qadM9u) {SB: link needs checking. Got link by clicking 'Create Invite' on the channel, but getting 'invalid link' msg for some reason when I try it locally). A channel node will be connected to all messages in that channel. 
```
Generated a non expiring link here: `https://discord.gg/32hXRee`

s-ben replied:
Used

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+[INSERT SCREENSHOT MESSAGE]
+
+- **Channel**:
```
So channels arent actually their own nodes in the graph, they are just a part of the address structure for messages and reactions. e.g. for a message its something like `sourcecred/discord/message/channelid/messageid`

Not sure if this will change in the future though, but yea I would probably just change the writing here to reflect that its not a node. 

Something valuable to add here might be to list the address structure for each node type, that way people can know how the channels come into play.

s-ben replied:
Valuable info if you're developing with the Discord plugin. Going to remove it for now though, as we're just listing actual nodes here and I want to get something from the user perspective up asap.

s-ben replied:
I think this doc is largely going to be used by people installing and configuring the plugin. Not as much developers.  And it's already getting quite long. Suggest saving this info for more dev-facing docs.

hammadj commented on docs/beta/plugins/discord.md:
````diff
@@ -1,5 +1,248 @@
+
+A Discord channel, e.g. [#start-here](https://discord.gg/qadM9u) {SB: link needs checking. Got link by clicking 'Create Invite' on the channel, but getting 'invalid link' msg for some reason when I try it locally). A channel node will be connected to all messages in that channel. 
+
+Found this in the code and pasting here in case the plugin exposes these channel types. 
+
+```
+ * Discord Channels can be of various types, defined below.
+ * See: https://discordapp.com/developers/docs/resources/channel#channel-object-channel-types
+ */
+export type ChannelType =
+  | "GUILD_TEXT"
+  | "DM"
+  | "GUILD_VOICE"
+  | "GROUP_DM"
+  | "GUILD_CATEGORY"
+  | "GUILD_NEWS"
+  | "GUILD_STORE";
+```
````
Don't think this is relevant to the actual graph, its just used internally when getting the data.

s-ben replied:
Removed.

s-ben replied:
Removed.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+An "authors" edge connects an author (user or bot) to a message they created. 
+
+- **References**: 
```
There's no references in the discord plugin currently, the closest thing would be "mentions" which is when you @ someone, it creates an edge from that message to the person you mentioned.

s-ben replied:
So is this a new edge type? Or is it a 'Reacts to' or 'Adds reaction', and I just need to update the definitions of those?

hammadj commented on docs/beta/plugins/discord.md:
````diff
@@ -1,5 +1,248 @@
+References in Discourse can be either hyperlinks to referencable nodes, or quoted parts of other posts {SB: I'm kind of guessing here, would be good to get more detailed explanation, and perhaps include a illustrative example (e.g. a quote that links to both text and image if supported?). This brings up another questions I've had for a while: when you select and click the "Quote" button in the UI ...
+```
+
+- **Reply to**:
````
Doesn't exist in the current implementation

s-ben replied:
Removed.

s-ben replied:
Removed.

hammadj commented on docs/beta/plugins/discord.md:
````diff
@@ -1,5 +1,248 @@
+The Discourse plugin creates the following kinds of edges (connections between nodes) 
+in the contribution graph:
+```
+- **Creates Reaction**:
````
There's two here technically, "addsReaction" and "reactsTo".

"Adds reaction" is an edge from the person who reacted to the reaction

"Reacts to" is an edge from that reaction to the message its on.

![image](https://user-images.githubusercontent.com/7143583/91104620-5b6be000-e62b-11ea-92fa-a54575eac020.png)

s-ben replied:
I think I got it. How's this?

- **Reacts to**:

A "Reacts to" edge connects a reaction (emoji) to the message reacted to. 

- **addsReaction**:

A "Adds Reaction" edge connects a reaction (emoji) to the author (user) of the reaction.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+{SB: the below para wasn't in the original docs, but is just me speculating based
+on past convos. We may want to remove if we can't provide enough
+guideance for it to be useful}
+Another caveat is that when a Discourse forum is very new, and does not have 
```
Some references to Discourse here. Was this meant to stay?

s-ben replied:
Not meant to stay. Just to give a reference for whoever writes the Status and Caveats section for Discord. Do you have any issues in mind that a user should know here? More easily gamable than other plugins? Known bugs? Design quirks?

s-ben replied:
Nah. Just left for reference/inspiration to whoever wanted to write the Status and Caveats. Have removed.

hammadj commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+A "creates reaction" edge connects a reaction node to its author and reacted to message. {SB: just guessing here}
+
+## Configuration
```
This is solid, but I'm wary about having two sources of truth for these instructions. There's also similar setup instructions in the template-instance repo. I feel like any content that is duplicated there should be removed from this doc and replaced with a link to the template-instance README, but we can keep some of the more specific extra stuff like custom emojis here and have the template-instance README link to this doc for more specific instructions.

s-ben replied:
So the whole configuration section is basically a rewrite of the `template-instance` repo README. Was actually wondering how best to handle this from a Cred perspective, and thought about adding you as a co-author on the PR, but figured you've already got Cred from template-instance? Let me know if that's not right though or you have other ideas. 

I actually think we should keep this content here in the docs, remove it from template-instance. We have this beautiful docs site now, and most of this is user-facing (if technical) documentation. This type of doco generally lives in a docs site if one exists. GitHub is mostly used for developer-facing docs. Typically dev docs are on the docs site too, unless there's no proper place to put them (separate dev docs site or section). Open to changing my mind though:)

s-ben commented on docs/beta/plugins/discord.md:
```diff
@@ -1,5 +1,248 @@
+
+- **Message**:
+
+A Discord message, e.g. ["Welcome to the SourceCred community!..."](https://discordapp.com/channels/453243919774253079/715770410955964477/715776215528505385) {SB: this is a link to the welcome message in the SourceCred Discord #start-here channel}. A message node will be connected to its author (user node creating the message), emoji reactions to the message, and any references to the message {SB ...
```
Got it.

s-ben commented:
@hammadj, addressed all your comments. Good stuff, I think we're almost there. Just have a couple outstanding questions/suggestions I think it would be good to address before publishing.
